design-toolkit/components/src/components/tree-view/index.tsx

```diff
@@ -0,0 +1,308 @@
+    features: [
+      syncDataLoaderFeature,
+      selectionFeature,
+      hotkeysCoreFeature,
+      dragAndDropFeature,
+      keyboardDragAndDropFeature,
+    ],
+  });
+
+  if (!items) {
+    return null;
+  }
+
+  return (
+    <TreeContext.Provider
+      value={{ variant, selectionMode, allowsDragging, showRuleLines }}
+    >
+      <div
+        {...tree.getContainerProps()}
+        className='fg-default-light overflow-x flex w-full flex-col outline-hidden'
+      >
+        {typeof children === 'function'
+          ? tree.getItems().map((item) => {
+              const lastOfSet =
+                item.getItemMeta().setSize - 1 === item.getItemMeta().posInSet;
+              return (
+                <div key={item.getId()} data-last-of-set={lastOfSet}>
+                  {children({ item, variant })}
+                </div>
+              );
+            })
+          : children}
+        <div
+          className='absolute border border-highlight-hover'
+          style={{ ...tree.getDragLineStyle() }}
+        />
+      </div>
+    </TreeContext.Provider>
+  );
+}
+
+TreeView.displayName = 'TreeView';
+
+const treeNodeStyles = cva(
+  'flex items-center border-transparent border-t border-b hover:bg-interactive-hover-dark',
+  {
+    variants: {
+      variant: {
+        cozy: 'icon-size-xl gap-s text-header-m',
+        compact: 'icon-size-l gap-s text-header-s', //icons
+        tight: 'icon-size-l gap-xxs text-header-s',
+      },
+      isSelected: {
+        false: 'fg-default-dark',
+      },
+      isReadOnly: {
+        true: 'fg-default-dark',
+      },
+      isDragTarget: {
+        true: 'bg-interactive-hover-dark',
+      },
+    },
+  },
+);
+
+export interface TreeNodeProps<T> extends VariantProps<typeof treeNodeStyles> {
+  children: ReactNode;
+  item: ItemInstance<T>;
+}
+
+function SelectionIcons({
+  selectionMode,
+  isSelected,
+}: { selectionMode: TreeSelectionMode; isSelected: boolean }) {
+  if (selectionMode === 'visibility') {
+    return <Icon>{isSelected ? <Show /> : <Hide />}</Icon>;
+  }
+
+  if (selectionMode === 'checkbox') {
+    return (
+      <Icon>{isSelected ? <CheckboxSelected /> : <CheckboxUnselected />}</Icon>
+    );
+  }
+
+  // selection === 'none'
+  return null;
+}
+
+function TreeNode<T extends TreeNodeType>({
+  children,
+  item,
+}: TreeNodeProps<T>) {
+  const { variant, selectionMode, allowsDragging, showRuleLines } =
+    useContext(TreeContext);
+
+  const {
+    getProps,
+    toggleSelect,
+    isSelected,
+    isFolder,
+    expand,
+    collapse,
+    isExpanded,
+  } = item;
+
+  const { onClick, ...props } = getProps();
+
+  const expander = isExpanded() ? (
+    <button type='button' onClick={expand} {...getProps()}>
+      <Icon>
+        <ChevronUp />
+      </Icon>
+    </button>
+  ) : (
+    <button type='button' onClick={collapse} {...getProps()}>
+      <Icon>
+        <ChevronDown />
+      </Icon>
+    </button>
+  );
+
+  const spacer = (
```

Comment:
Perf nit: hoist this out of this component since it does not use any dynamic props or state